### PR TITLE
Give starter.yaml some love

### PR DESCRIPTION
**Pull request by alvaroaleman**
* Change the namespace for Prow to `prow`, the `default` namespace
  shouldn't get polluted
* Add and use `ghproxy`, it became a requirement recently
* Rename the ingress from `ing` to `prow` to make it more meaningful
* Stop using `NodePort` services when we only need clusterIP, thats
  useless at best and a security issue at worst
* Use `github-token` and not `oauth-token` as secret name for the github
  token, the latter describes a technology, not a purpose, which is
  confusing
* Add templates for github-token and hmac-token secrets
* Set `decorate_all_jobs: true`, there is no reason to not use
  decoration for new deployments
* Make Ingress compatible with cert-manager
* Enable spyglass, Gubernator is deprecated
* Use minio, so no external object storage needs to be set up
* Add plugins config
* Add spyglass config
* Enable Inrepoconfig
* Add Tide query
* Fix Deck RBAC
* Use prow-controller-manager/plankv2
* Update decoration-utils images
* Update references to `starter.yaml`

Basically: Bring the thing up-to-date with how we expect Prow to be set up and make it something that "just works" after replacing a couple of variables rather than "Do these five manual steps, deploy `starter.yaml`, do another five manual steps.

/assign @spiffxp @fejta @sbueringer 
/cc @saschagrunert 
as you were interested in using prow in an on-prem scenario/without cloud storage.

In the future we could probably make this part of the `stable` promotion process by building it out to be used in a kind-based e2e test that uses something like `ngrok` to expose the hook endpoint.

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
/lgtm
/hold

These changes look good to me, nice to see this modernized.

File: config/prow/cluster/starter.yaml [hunk 1]
```diff
@@ -3,24 +3,149 @@
...
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: prow
+  name: s3-credentials
+stringData:
+  service-account.json: |
+    {
+      "region": "minio",
+      "access_key": "<<CHANGE_ME_MINIO_ACCESS_KEY>>",
```
Comment (fejta) [inline config/prow/cluster/starter.yaml:60]: It doesn't seem like we should be encouraging people to include secrets in the starter.yaml file.

File: config/prow/cluster/starter.yaml [hunk 2]
```diff
@@ -876,3 +923,400 @@ roleRef:
...
+        - /data/prow-logs
+        volumeMounts:
+        - name: data
+          mountPath: "/data"
+      containers:
+      - name: minio
+        volumeMounts:
+        - name: data
+          mountPath: "/data"
+        image: minio/minio:latest
```
Comment (fejta) [inline config/prow/cluster/starter.yaml:1287]: Shouldn't we pick a specific version that we've tested against?

**Comment by alvaroaleman** [comment]
> A lot of these changes seem overly coupled to S3. This unnecessarily increases the length/complexity of the file for people on GCP.

No. They are coupled to an object storage that can be self-hosted which happens to have a S3-compatible API as there is no such thing with a GCS-compatible API (to my knowledge). The overall goal of the `starter.yaml` file is to allow ppl to check out Prow and make it as easy as possible to get to the point of having a working Prow setup with as little work as possible.

Setting up object storage on any cloud provider is par definition always more work than applying a couple of manifests. Most ppl that stick with Prow will want that, which is why the `prow/getting_started_deploy.md` contains a section on how to do that.

File: config/prow/cluster/starter.yaml (hunk 1 above)
Comment (alvaroaleman) [reply config/prow/cluster/starter.yaml:60]: Its not particularly nice but the thing is not publicly exposed, hence its not really a big deal IMHO

File: config/prow/cluster/starter.yaml (hunk 2 above)
Comment (alvaroaleman) [reply config/prow/cluster/starter.yaml:1287]: But then we have to regularly update it. The S3 api the thing implements can be assumed to be very stable.

**Comment by fejta** [comment]
The issue I have is that all the minio volumes, secrets, deployments, etc are totally unnecessary. And we should make it easy to delete them. `kubectl delete -f minio.yaml` accomplishes that. And `kubectl create -f starter.yaml -f minio.yaml` is not substantially harder than when there is just one `-f` flag.

File: config/prow/cluster/starter.yaml (hunk 1 above)
Comment (sbueringer) [reply config/prow/cluster/starter.yaml:60]: Another option could be to use some kind of envsubst magic. But no idea if we want that/ if it's better.

File: config/prow/cluster/starter.yaml [hunk 3]
```diff
@@ -3,24 +3,149 @@
...
+    prowjob_namespace: prow
     pod_namespace: test-pods
+
+    in_repo_config:
+      enabled:
+        "*": true
+
+    deck:
+     spyglass:
+       lenses:
```
Comment (sbueringer) [inline config/prow/cluster/starter.yaml:99]: Just fyi. If we have a publicly accessible Minio URL this would also work:
gcs_browser_prefix: <minio-url>/minio/

But not sure if it's overly complicated for the starter.yaml

**Comment by sbueringer** [comment]
@fejta @alvaroaleman I think if we want to decouple it from s3, it makes sense to use e.g. kustomize. Because the s3 part is not only deploying Minio but also different flags in deck etc.

File: config/prow/cluster/starter.yaml [hunk 4]
```diff
@@ -3,24 +3,149 @@
...
+      queries:
+      - labels:
+        - lgtm
+        - approved
+        missingLabels:
+        - needs-rebase
+        - do-not-merge/hold
+        - do-not-merge/work-in-progress
+        - do-not-merge/invalid-owners-file
+        orgs:
```
Comment (sbueringer) [inline config/prow/cluster/starter.yaml:143]: I'm also setting:
```
pr_status_base_urls:
    '*': https://prow.<< your-domain.com >>/pr
```

File: config/prow/cluster/starter.yaml [hunk 5]
```diff
@@ -3,24 +3,149 @@
...
+           name: junit
+         required_files:
+         - .*/junit.*\.xml
+       - lens:
+           name: podinfo
+         required_files:
+         - podinfo.json
+
+    plank:
+      job_url_prefix_config:
```
Comment (sbueringer) [inline config/prow/cluster/starter.yaml:119]: There's also `job_url_template` and `report_templates` were I included a template with the prow domain in my deployment. Not sure if it's important here

File: config/prow/cluster/starter.yaml [hunk 6]
```diff
@@ -146,14 +271,17 @@ spec:
         args:
         - --dry-run=false
         - --config-path=/etc/config/config.yaml
```
Comment (sbueringer) [inline config/prow/cluster/starter.yaml:273]: Sorry it's not about fixing my deployment but don't we need `job-config-path` and the mounted jobs in hook?

Comment (sbueringer) [reply config/prow/cluster/starter.yaml:273]: Ah I guess we dont need it because we use in-repo config?

File: config/prow/cluster/starter.yaml [hunk 7]
```diff
@@ -420,48 +516,57 @@ spec:
...
+        - --github-token-path=/etc/github/token
+        - --github-endpoint=http://ghproxy
+        - --github-endpoint=https://api.github.com
+        - --github-graphql-endpoint=http://ghproxy/graphql
```
Comment (sbueringer) [inline config/prow/cluster/starter.yaml:522]: Probably not necessary for starter:
```
        - --status-path=s3://tide/tide-status
        - --history-uri=s3://tide/tide-history.json
```

File: config/prow/cluster/starter.yaml [hunk 8]
```diff
@@ -499,9 +608,11 @@ spec:
         - --continue-on-error=true
         - --plugin-config=/etc/plugins/plugins.yaml
         - --config-path=/etc/config/config.yaml
-        - --github-token-path=/etc/github/oauth
+        - --github-token-path=/etc/github/token
+        - --github-endpoint=http://ghproxy
+        - --github-endpoint=https://api.github.com
```
Comment (sbueringer) [inline config/prow/cluster/starter.yaml:613]: Probably not necessary for starter:
```
- --status-path=s3://status-reconciler/status-reconciler-status
```

File: config/prow/cluster/starter.yaml (hunk 5 above)
Comment (alvaroaleman) [reply config/prow/cluster/starter.yaml:119]: So `job_url_template` is ignored when `plank.GetJobURLPrefix(pj.Spec.Refs) != ""` and should just not be used.
I'll add `report_template`. We are so bad at defaulting.

File: config/prow/cluster/starter.yaml (hunk 6 above)
Comment (alvaroaleman) [reply config/prow/cluster/starter.yaml:273]: Its a single file that has both prow config and job config

File: config/prow/cluster/starter.yaml (hunk 3 above)
Comment (alvaroaleman) [reply config/prow/cluster/starter.yaml:99]: In this sample the minio is not public

File: config/prow/cluster/starter.yaml (hunk 4 above)
Comment (alvaroaleman) [reply config/prow/cluster/starter.yaml:143]: Yeah but this just gives an empty, non-erroring page unless you set up a github app

**Comment by alvaroaleman** [comment]
> I think if we want to decouple it from s3, it makes sense to use e.g. kustomize. Because the s3 part is not only deploying Minio but also different flags in deck etc.

I really hate templating as its usually the super important glue that doesn't have any kind of testing. But yeah, I guess that has to be done

File: config/prow/cluster/starter.yaml (hunk 4 above)
Comment (sbueringer) [reply config/prow/cluster/starter.yaml:143]: Ah right, makes sense

File: config/prow/cluster/starter.yaml (hunk 2 above)
Comment (sbueringer) [reply config/prow/cluster/starter.yaml:1287]: Should be pretty stable with Minio. We're using it since about over 2 years with any problems in this regard.

File: config/prow/cluster/starter.yaml (hunk 5 above)
Comment (sbueringer) [reply config/prow/cluster/starter.yaml:119]: Good to know..

File: config/prow/cluster/starter.yaml (hunk 6 above)
Comment (sbueringer) [reply config/prow/cluster/starter.yaml:273]: kk got it

File: config/prow/cluster/starter.yaml [hunk 9]
```diff
@@ -3,24 +3,149 @@
 # their own files.
```
Comment (tseelbach) [inline config/prow/cluster/starter.yaml:3]: perhaps add to line 2/3 comment.  Something like:
"For examples see the per component yaml files in this directory."

**Comment by alvaroaleman** [comment]
@fejta @sbueringer PTAL. Ive played around a bit with using kustomize, but its super finicky for our use case because we have to do things like "Adjust the embedded yaml" or "add this arg to that deployment". So I ended up just creating two files, one for minio/s3 and one for gcs.

I can create a follow-up that adds a script that checks the diff of the two files against a known-accepted diff to prevent drift. Follow-up because this PR collects a merge conflict every time prow gets bumped which is a bit annoying.

File: config/prow/cluster/starter-gcs.yaml [hunk 10]
```diff
@@ -3,24 +3,123 @@
...
+        "*": https://prow.<< your-domain.com >>/view/gcs
+      report_templates:
+        '*': >-
+            [Full PR test history](https://prow..<< your-domain.com >>/pr-history?org={{.Spec.Refs.Org}}&repo={{.Spec.Refs.Repo}}&pr={{with index .Spec.Refs.Pulls 0}}{{.Number}}{{end}}).
+            [Your PR dashboard](https://prow.<< your-domain.com >>/pr?query=is:pr+state:open+author:{{with
+            index .Spec.Refs.Pulls 0}}{{.Author}}{{end}}).
+      default_decoration_configs:
+        "*":
+          gcs_configuration:
+            bucket: gcs://bucket-name
```
Comment (sbueringer) [inline config/prow/cluster/starter-gcs.yaml:98]: Should be `gs://` (same in other cases)

File: config/prow/cluster/starter-gcs.yaml [hunk 11]
```diff
@@ -420,48 +489,60 @@ spec:
         args:
         - --dry-run=false
         - --config-path=/etc/config/config.yaml
+        - --github-token-path=/etc/github/token
+        - --github-endpoint=http://ghproxy
+        - --github-endpoint=https://api.github.com
+        - --github-graphql-endpoint=http://ghproxy/graphql
+        - --gcs-credentials-file=/etc/gcs-credentials/service-account.json
```
Comment (sbueringer) [inline config/prow/cluster/starter-gcs.yaml:496]: mount point for the gcs credentials is missing

File: config/prow/cluster/starter-gcs.yaml [hunk 12]
```diff
@@ -499,9 +584,13 @@ spec:
         - --continue-on-error=true
         - --plugin-config=/etc/plugins/plugins.yaml
         - --config-path=/etc/config/config.yaml
-        - --github-token-path=/etc/github/oauth
+        - --github-token-path=/etc/github/token
+        - --github-endpoint=http://ghproxy
+        - --github-endpoint=https://api.github.com
+        - --gcs-credentials-file=/etc/gcs-credentials/service-account.json
```
Comment (sbueringer) [inline config/prow/cluster/starter-gcs.yaml:590]: mount point is missing

**Comment by sbueringer** [comment]
@alvaroaleman A few nits. What do you think about calling the secret for s3 and gcs just something generic like `blob-storage-credential`. Not sure if it's worth it, but then there would almost be no diff between both deployments (only a few bucket paths, gcs-credentials-file vs s3-credentials file and the Minio depoyment I guess)

File: config/prow/cluster/starter-s3.yaml [hunk 13]
```diff
@@ -0,0 +1,1346 @@
...
+  service-account.json: |
+    {
+      "region": "minio",
+      "access_key": "<<CHANGE_ME_MINIO_ACCESS_KEY>>",
+      "endpoint": "minio.prow.svc.cluster.local",
+      "insecure": true,
+      "s3_force_path_style": true,
+      "secret_key": "<<CHANGE_ME_MINIO_SECRET_KEY>>"
+    }
+---
```
Comment (sbueringer) [inline config/prow/cluster/starter-s3.yaml:1273]: Should we add similar dummy credentials in the gcs yaml?

**Comment by stevekuznetsov** [comment]
@enj

File: config/prow/cluster/starter.yaml (hunk 1 above)
Comment (tseelbach) [reply config/prow/cluster/starter.yaml]: About secrets in starter.yaml:
- we have enterprise rules that we cannot commit secrets/passwords into git repos. 
- as a new Prow user, the new starter.yaml worked almost perfectly(Thank you!)  
- So following gitOps approach I wanted to check it in, so my team could start customizing, and so i could roll back when we mess up (remember we are new) But first i had to split out the secret creation.    It would have been better to have secret creation as a separate step in the first place.

Comment (alvaroaleman) [reply config/prow/cluster/starter.yaml]: I disagree. The purpose of this is to get ASAP something that works so ppl can start toying around and check if Prow fits their requirements. If and how ppl solve the problem of not having this in their yaml is not something that is supposed to get solved here. Removing a secret from a yaml manifest is trivial. Adding it and with it the correct keys might not be as trivial.

File: config/prow/cluster/starter-s3.yaml (hunk 13 above)
Comment (alvaroaleman) [reply config/prow/cluster/starter-s3.yaml:1273]: IMHO its less relevant because you can just download a working credential from GCP whereas what we have for AWS is a custom format

**Comment by alvaroaleman** [comment]
@fejta @sbueringer rebased and existing feedback addressed. Even if not perfect, its IMHO a lot better to get someone up and running quickly than what we had before. Since this is a rebase magnet, any chance we can merge it and do follow-ups if we find more issues?

**Comment by sbueringer** [comment]
@alvaroaleman  All findings from my side addressed so 
/lgtm 

anyway :)

**Comment by sbueringer** [comment]
/lgtm 
:)
